polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -173,42 +173,54 @@ class LoopGuardBuilder {
 public:
   static std::unique_ptr<LoopGuardBuilder> create(LoopLikeOpInterface loop);
 
-  LoopGuardBuilder(LoopLikeOpInterface loop) : builder(loop){};
+  LoopGuardBuilder(LoopLikeOpInterface loop) : builder(loop), loop(loop){};
   LoopGuardBuilder(const LoopGuardBuilder &) = delete;
   LoopGuardBuilder(LoopGuardBuilder &&) = delete;
   void operator=(const LoopGuardBuilder &) = delete;
   void operator=(LoopGuardBuilder &&) = delete;
   virtual ~LoopGuardBuilder() = default;
 
-  virtual void guardLoop() const = 0;
+  void guardLoop() const;
 
 protected:
+  static Block *getThenBlock(RegionBranchOpInterface ifOp) {
+    return &*ifOp->getRegion(0).begin();
+  }
+  static Block *getElseBlock(RegionBranchOpInterface ifOp) {
+    return &*ifOp->getRegion(1).begin();
+  }
+  virtual Operation::operand_range getInitVals() const = 0;
+
   mutable OpBuilder builder;
+
+private:
+  void replaceUsesOfLoopReturnValues(RegionBranchOpInterface &) const;
+  virtual RegionBranchOpInterface createGuard() const = 0;
+
+  mutable LoopLikeOpInterface loop;
 };
 
 class SCFLoopGuardBuilder : public LoopGuardBuilder {
 public:
-  SCFLoopGuardBuilder(LoopLikeOpInterface loop) : LoopGuardBuilder(loop) {}
-
-protected:
-  scf::IfOp guard(LoopLikeOpInterface loop) const;
+  SCFLoopGuardBuilder(LoopLikeOpInterface loop)
+      : LoopGuardBuilder(loop), loop(loop) {}
 
 private:
+  RegionBranchOpInterface createGuard() const final;
   virtual Value createGuardExpr() const = 0;
-  virtual scf::IfOp createGuard() const = 0;
-  virtual void replaceUsesOfLoopReturnValues(scf::IfOp &) const {}
+
+  mutable LoopLikeOpInterface loop;
```

Comment:
The base class (LoopGuardBuilder) already has this member so is not needed here.